lib/src/rules/avoid_relative_imports.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2018, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+import '../ast.dart';
+
+const _desc = r'Avoid relative imports for all files`.';
+
+const _details = r'''*DO* avoid relative imports for files in `lib/`.
+
+When mixing relative and absolute imports it's possible to create confusion
+where the same member gets imported in two different ways. One way to avoid
+that is to ensure you consistently use absolute imports for files withing the
+`lib/` directory.
+
+This is the opposite of 'prefer_relative_imports'.
+
+**GOOD:**
+
+```
+import 'package:foo/bar.dart';
+
+import 'package:foo/baz.dart';
+
+import 'package:foo/src/baz.dart';
+...
+```
+
+**BAD:**
+
+```
+import 'baz.dart';
+
+import 'src/bag.dart'
+
+import '../lib/baz.dart';
+
+...
+```
+
+''';
+
+class AvoidRelativeImports extends LintRule implements NodeLintRule {
+  AvoidRelativeImports()
+      : super(
+            name: 'avoid_relative_imports',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+    registry.addImportDirective(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+  final LinterContext context;
+
+  bool isInLibFolder;
+
+  _Visitor(this.rule, this.context);
+
+  bool isRelativeImport(ImportDirective node) {
+    // Ignore this compilation unit if it's not in the lib/ folder.
+    if (!isInLibFolder) return false;
```

Comment:
I believe that this PR has the opposite problem from the previous PR in that this will have false negatives in the case where a library outside of `lib` is using a relative path to reference a library inside of `lib` (which should also be disallowed).

Reply:
I added some clarification in description:
```
const _desc = r'Avoid relative imports for files in `lib/`.';
```
and details:
```
Might be used with 'avoid_relative_lib_imports' to avoid relative imports of
files within `lib/` directory outside of it. (for example `test/`)
```


It seems strange to duplicate logic of another lint rule, so I added a link to lint that solves problem for files outside of `lib/`.


Does this seem reasonable to you?